src/sss_client/ssh/sss_ssh_knownhosts.c

```diff
@@ -0,0 +1,210 @@
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <stdio.h>
+#include <talloc.h>
+#include <popt.h>
+#include <signal.h>
+#include <netdb.h>
+
+#include "util/util.h"
+#include "util/sss_ssh.h"
+#include "sss_client/sss_cli.h"
+#include "sss_client/ssh/sss_ssh_client.h"
+
+static errno_t known_hosts(TALLOC_CTX *mem_ctx, const char *domain, const char *host,
+                           struct sss_ssh_ent **_ent, const char **_keyhost)
+{
+    TALLOC_CTX *tmp_ctx;
+    errno_t ret;
+    struct addrinfo ai_hint;
+    struct addrinfo *ai = NULL;
+    char canonhost[NI_MAXHOST];
+    const char *canonname = NULL;
+    const char *keyhost;
+    struct sss_ssh_ent *ent = NULL;
+
+    if (_ent == NULL || _keyhost == NULL) {
+        DEBUG(SSSDBG_CRIT_FAILURE, "Invalid NULL received\n");
+        ERROR("Internal error\n");
+        return EINVAL;
+    }
+
+    tmp_ctx = talloc_new(NULL);
+    if (tmp_ctx == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE, "Failed to allocate tmp_ctx\n");
+        ERROR("Not enough memory\n");
+        return ENOMEM;
+    }
+
+    memset(&ai_hint, 0, sizeof(struct addrinfo));
+    ai_hint.ai_family = AF_UNSPEC;
+    ai_hint.ai_socktype = SOCK_STREAM;
+    ai_hint.ai_protocol = IPPROTO_TCP;
+    ai_hint.ai_flags = AI_NUMERICHOST;
+
+    DEBUG(SSSDBG_FUNC_DATA, "Looking up canonical name for: %s\n", host);
+    ret = getaddrinfo(host, NULL, &ai_hint, &ai);
+    if (ret != EOK) {
+        ai_hint.ai_flags = AI_CANONNAME;
+        ret = getaddrinfo(host, NULL, &ai_hint, &ai);
+        if (ret != EOK) {
+            DEBUG(SSSDBG_OP_FAILURE,
+                  "getaddrinfo() failed (%d): %s\n", ret, gai_strerror(ret));
+            goto done;
+        } else {
+            canonname = ai->ai_canonname;
+        }
+    } else {
+        ret = getnameinfo(ai->ai_addr, ai->ai_addrlen,
+                          canonhost, NI_MAXHOST, NULL, 0, NI_NAMEREQD);
+        if (ret != EOK) {
+            DEBUG(SSSDBG_OP_FAILURE,
+                  "getnameinfo() failed (%d): %s\n", ret, gai_strerror(ret));
+            goto done;
+        } else {
+            canonname = canonhost;
+        }
+    }
+    DEBUG(SSSDBG_FUNC_DATA, "Found canonical name: %s\n", canonname);
+
+    /* look up public keys */
+    ret = sss_ssh_get_ent(tmp_ctx, SSS_SSH_GET_HOST_PUBKEYS,
+                          canonname, domain, host, &ent);
+    if (ret == ENOENT) {
+        DEBUG(SSSDBG_FUNC_DATA,
+              "sss_ssh_get_ent() found no entry\n");
+        goto done;
+    } else if (ret != EOK) {
+        DEBUG(SSSDBG_OP_FAILURE,
+              "sss_ssh_get_ent() failed (%d): %s\n", ret, sss_strerror(ret));
+        goto done;
+    }
+
+    keyhost = talloc_asprintf(tmp_ctx, "%s,%s", canonname, host);
```

Comment:
It will write "client.ipa.test,client.ipa.test" if host is already in canonical format. Can you make sure it is written only once?

Reply:
You are right it will right the name twice. This is not a problem as ssh handles this situation (sometimes the files end up with duplicated names, like in this case).
Instead, trying to detect when both values are the same requires:
* comparing the strings each time and,
*  use separate format strings for each case (one or two values).
It think it is better to leave it like it is.

Reply:
Ok, as you wish.